src/codec.rs

```diff
@@ -782,36 +782,71 @@ impl<T: Encode> Encode for [T] {
 	}
 }
 
-/// Read an `u8` vector from the given input.
-pub(crate) fn read_vec_u8<I: Input>(input: &mut I, len: usize) -> Result<Vec<u8>, Error> {
+/// Create a `Vec<T>` by casting directly from a buffer of read `u8`s
+///
+/// `T` must encodes to its byte representation, size of `T` must be less or equal to
+/// [`MAX_PREALLOCATION`].
+pub(crate) fn read_vec_from_u8s<I, T>(input: &mut I, items_len: usize) -> Result<Vec<T>, Error>
+where
+	I: Input,
+	T: ToMutByteSlice + Default + Clone,
+{
+	debug_assert!(MAX_PREALLOCATION >= mem::size_of::<T>(), "Invalid precondition");
+
+	let byte_len = items_len.checked_mul(mem::size_of::<T>())
+		.ok_or_else(|| "Item is too big and cannot be allocated")?;
+
 	let input_len = input.remaining_len()?;
 
 	// If there is input len and it cannot be pre-allocated then return directly.
-	if input_len.map(|l| l < len).unwrap_or(false) {
+	if input_len.map(|l| l < byte_len).unwrap_or(false) {
 		return Err("Not enough data to decode vector".into())
 	}
 
+	// In both these branches we're going to be creating and resizing a Vec<T>,
+	// but casting it to a &mut [u8] for reading.
+
 	// Note: we checked that if input_len is some then it can preallocated.
-	let r = if input_len.is_some() || len < MAX_PREALLOCATION {
+	let r = if input_len.is_some() || byte_len < MAX_PREALLOCATION {
 		// Here we pre-allocate the whole buffer.
-		let mut r = vec![0; len];
-		input.read(&mut r)?;
+		let mut items: Vec<T> = vec![Default::default(); items_len];
+		let mut bytes_slice = items.as_mut_byte_slice();
+		input.read(&mut bytes_slice)?;
 
-		r
+		items
 	} else {
+		// An allowed number of preallocated item.
+		// Note: `MAX_PREALLOCATION` is expected to be more or equal to size of `T`, precondition.
+		let max_preallocated_items = MAX_PREALLOCATION / mem::size_of::<T>();
+
 		// Here we pre-allocate only the maximum pre-allocation
-		let mut r = vec![];
-
-		let mut remains = len;
-		while remains != 0 {
-			let len_read = MAX_PREALLOCATION.min(remains);
-			let len_filled = r.len();
-			r.resize(len_filled + len_read, 0);
-			input.read(&mut r[len_filled..])?;
-			remains -= len_read;
+		let mut items: Vec<T> = vec![];
+
+		let mut items_remains = items_len;
+
+		while items_remains != 0 {
+			let items_len_read = max_preallocated_items.min(items_remains);
+
+			let items_len_filled = items.len();
+			let items_new_size = items_len_filled + items_len_read;
+
+			// This call is here to satisfy the existing
+			// vec_decode_right_capacity test.
+			// It's not obviously functionally necessary,
+			// nor is it obviously a problem.
+			items.reserve_exact(items_len_read);
+
+			items.resize_with(items_new_size, Default::default);
```

Comment:
Could you run the benchmarks of this against using `set_len` here?

Reply:
there is quite a difference it goes roughly from 5,7 to 4,5

I changed for `set_len`